src/sql/workbench/services/objectExplorer/browser/asyncServerTreeDragAndDrop.ts

```diff
@@ -0,0 +1,124 @@
+import { ServerTreeDragAndDrop } from 'sql/workbench/services/objectExplorer/browser/dragAndDropController';
+import { IDragAndDrop } from 'vs/base/parts/tree/browser/tree';
+import { ServerTreeElement } from 'sql/workbench/services/objectExplorer/browser/asyncServerTree';
+
+/**
+ * Implements drag and drop for the server tree
+ */
+export class AsyncServerTreeDragAndDrop implements ITreeDragAndDrop<ServerTreeElement> {
+
+	private _dragAndDrop: IDragAndDrop;
+
+	constructor(
+		@IConnectionManagementService connectionManagementService: IConnectionManagementService,
+	) {
+		this._dragAndDrop = new ServerTreeDragAndDrop(connectionManagementService);
+	}
+
+	/**
+	 * Returns a uri if the given element should be allowed to drag.
+	 * Returns null, otherwise.
+	 */
+	public getDragURI(element: ServerTreeElement): string {
+		return this._dragAndDrop.getDragURI(undefined, element);
+	}
+
+	/**
+	 * Returns a label(name) to display when dragging the element.
+	 */
+	public getDragLabel(elements: ServerTreeElement[]): string {
+		return this._dragAndDrop.getDragLabel(undefined, elements);
+	}
+
+	/**
+	 * Called when the drag operation starts.
+	 */
+	public onDragStart(dragAndDropData: IDragAndDropData, originalEvent: DragEvent): void {
+		// Force the event cast while in preview - we don't use any of the mouse properties on the
+		// implementation so this is fine for now
+		return this._dragAndDrop.onDragStart(undefined, dragAndDropData, <any>originalEvent);
+	}
+
+	public onDragOver(data: IDragAndDropData, targetElement: ServerTreeElement, targetIndex: number, originalEvent: DragEvent): boolean | ITreeDragOverReaction {
+		// Force the event cast while in preview - we don't use any of the mouse properties on the
+		// implementation so this is fine for now
+		const canDragOver = this._dragAndDrop.onDragOver(undefined, data, targetElement, <any>originalEvent);
+
+		if (canDragOver.accept) {
+			return TreeDragOverReactions.acceptBubbleDown(true);
+		} else {
+			return { accept: false };
+		}
+	}
+
+	/**
+	 * Handle a drop in the server tree.
+	 */
+	public drop(data: IDragAndDropData, targetElement: ServerTreeElement, targetIndex: number, originalEvent: DragEvent): void {
+		// Force the event cast while in preview - we don't use any of the mouse properties on the
+		// implementation so this is fine for now
+
+		// TODO: chgagnon Drop on root node
+		this._dragAndDrop.drop(undefined, data, targetElement, <any>originalEvent);
+	}
+
+	public onDragEnd(originalEvent: DragEvent): void {
+		TreeUpdateUtils.isInDragAndDrop = false;
+	}
+}
+
+export class AsyncRecentConnectionsDragAndDrop implements ITreeDragAndDrop<ServerTreeElement> {
+
+	/**
+	 * Returns a uri if the given element should be allowed to drag.
+	 * Returns null, otherwise.
+	 */
+	public getDragURI(element: ServerTreeElement): string | null {
+		if (element instanceof ConnectionProfile) {
+			return (<ConnectionProfile>element).id;
```

Comment:
`(element as ConnectionProfile).id`

Reply:
This is actually unnecessary due to the type guard - the compiler already knows it's a ConnectionProfile so no need to cast.